Full rewrite

rv32ima opened the pull request:
This is a PR tracking my status on rewriting Tarmac. Most of the code here will get touched - frankly a lot of it is inadequate, and not up to modern standards. I'm planning on keeping the user-facing API the same, just simplifying a lot of the code.

LPGhatguy committed:
Idea'

LPGhatguy committed:
Name ideas

LPGhatguy committed:
Hello, Tarmac

LPGhatguy committed:
Fill out Cargo metadata

LPGhatguy committed:
Throw in a bunch of gunk

LPGhatguy committed:
Pack together learnings from bash scripts into tool to upload images

LPGhatguy committed:
Add explicit auth, usage in README

LPGhatguy committed:
Add 'sync' command

LPGhatguy committed:
Add more feedback for uploads

LPGhatguy committed:
Update README

LPGhatguy committed:
Support multiple paths for sync command

LPGhatguy committed:
Handle malformed API responses a little better

LPGhatguy committed:
Expand and refactor configuration for the future

LPGhatguy committed:
Refactor 'sync' command

LPGhatguy committed:
Refactor commands into 'commands' mod folder

LPGhatguy committed:
Append 'Options' to option structs

LPGhatguy committed:
Add config method using snafu crate

LPGhatguy committed:
Add examples and slim down scope of config

LPGhatguy committed:
Write sync command v2

LPGhatguy committed:
Update README with new stuff

LPGhatguy committed:
Add support for codegen

LPGhatguy committed:
Add uploaded_subslice field to asset manifests for future spritesheets

LPGhatguy committed:
Add 'slice' codegen prototype

LPGhatguy committed:
Add stub to generate spritesheets and coalesce assets in the future

LPGhatguy committed:
Remove unnecessary passing of root_path in SyncSession

LPGhatguy committed:
Introduce new AssetName type instead of just using String everywhere

ZoteTheMighty committed:
Remove extra '.' in file extensions for code-gen'd filename (#3)

LPGhatguy committed:
Release 0.1.0 to pave the way for new work

LPGhatguy committed:
Refactor Tarmac to think in terms of "groups" (#1)

* Experiment with refactoring to think about asset groups

* pub all the things

* Flesh out types a bit

* Reorganize new types

* Flesh out operations on new types

* Scrape out old sync process, start filling in new types

* Strip out more gunk

* Shuffle project/manifest finding into SyncSession::new

* Start shaping how inputs are gathered and turned back into manifests

* Actually generate manifests from inputs and groups

* Add GitHub Actions workflow

* Flesh out baseline syncing to Roblox

LPGhatguy committed:
Update version to 0.2.0-dev and update README

LPGhatguy committed:
Update some dependencies

LPGhatguy committed:
Refactor again (#4)

* Throw everything around with new ideas

* Trim unused imports

* Hoist name and max-spritesheet-size into Config and remove ProjectConfig

* Expose more details from Config type

* Spell out config discovery function

* Recurse through found projects

* Process inputs across configs (additions to refactor in progress) (#6)

* First pass

* Loop over input configs and gather all inputs

* Cleaned up input discovery a bit

* Stub out more sync command

* Update manifest format

* Clean up auth strategy in sync

* Fix filtering problems

* Replace unnecessary parent/unwrap

* Narrow inputs search by narrowing search area based on non-pattern prefix of glob

* Introduce a basic game as an example project

* Fix botched rebase introducing game project

* Ignore Studio place lock files

* Make more types PartialEq

* Update examples

* Throw down sync process roughly

* Tidy up placeholder comments in sync

* Move error type to bottom of file

* Improve comments/readability on sync process

* Rename SyncError to just Error, and re-export it as SyncError

* Make sync actually upload things

* Clean up sync more

* Add TODO comment for UploadResponse

* Incremental sync basics done

* Implement codegen for asset-url

* Run Tarmac on examples

* Delete old sync process

* Appease clippy

* Refactor configs to all be stored in Vec

* Make SyncSession work in terms of config indices

* Flatten config into manifest

* Undo using config index

* Generate new manifest for example #2

* First pass at packing images

* WIP spritesheet outputting

* Maxrects algo, honors max spritesheet size

* Clean up logic and add more comments

* Change AssetName to be use an Arc<str> instead of a String

* Add DPI scale detecting function

* Integrate DPI scale function into sync

* Compute hashes for pack inputs, abort early if no inputs have changed

* Adjust slice data to be more toml-friendly

* Sync spritesheet objects, populate manifest accordingly

* Move codegen into codegen.rs

* Make packable images in examples RGBA

* Codegen for slices

* Run Tarmac on example 2

* You're welcome, clippy

* Last change before completion

Co-authored-by: Paul Doyle <37384169+ZoteTheMighty@users.noreply.github.com>

LPGhatguy committed:
Update README

LPGhatguy committed:
Build with static CRT, add release workflow

LPGhatguy committed:
Release 0.2.0

LPGhatguy committed:
Add link to releases page for installation

LPGhatguy committed:
Fancier README

LPGhatguy committed:
Introduce better crash messaging, add verbosity flag

LPGhatguy committed:
Add basic debug target for debugging Tarmac

LPGhatguy committed:
Ignore .tarmac-debug debugging directory

LPGhatguy committed:
Use IDs for debug upload target

LPGhatguy committed:
Switch to new rectangle packing library (#12)

* Tinker with custom rectangle/spritesheet packing library

* Scaffold out packing algorithm a little

* Packing is functional

* Break into modules

* Expose more API, replace Sheep

* Placeholder blit function

* Add (maybe broken) image blitting

* Update example project, now using a square image with new packer

* Shear sheep

* Break out image stuff into a separate mod

* Fix the code

* Switch from SHA-256 to BLAKE3 for image hashing

* Add padding on bottom-right of images

* Remove debug checks in packos

* Rerun on example projects

* Fix failing examples

LPGhatguy committed:
Add release workflow for Linux

LPGhatguy committed:
Update packos docs

LPGhatguy committed:
Update Packos README

LPGhatguy committed:
Improve naming of items in Packos

LPGhatguy committed:
Improve Packos API and add documentation

LPGhatguy committed:
Update Packos README

LPGhatguy committed:
Release 0.3.0

LPGhatguy committed:
Update README to mention pre-built Linux binaries

LPGhatguy committed:
Move UploadStrategy out of sync.rs into sync_backend.rs

LPGhatguy committed:
Read input images in packing all at once

LPGhatguy committed:
Rework packed image pipeline (#19)

* Work on sync algorithm for packed images

* Read inputs and calculate hashes earlier in pipeline

* Make hash in manifest non-optional

* Traverse input groups in order via BTreeMap

* Sort inputs and refactor InputCompatibility to InputKind

* Fix incremental sync

LPGhatguy committed:
Release 0.3.1

LPGhatguy committed:
Add methods to get/set pixels in Image

LPGhatguy committed:
Alpha bleed the results of packed images

LPGhatguy committed:
Update Changelog

LPGhatguy committed:
Codegen v2 (#22)

* Some form of batch codegen

* Switch to building an AST

* Literal -> Expression, TableLiteral -> Table

* Add codegen-path config, move SyncInput into data

* Improve naming of sync functions

* Try to move codegen away from sync.rs

* Improve codegen docs and minor usage

* Update spritesheet example to use packed codegen, finding bugs

* Fix config paths being non-absolute and causing problems

* Switch codegen from HashMap to BTreeMap for deterministic ordering

* Pull in 'fs' layer from testez-cli and foreman

* Loosen Lua ident restrictions

* Fix rotted comment

* Generate paths in codegen from paths instead of names

* Improve docs and remove unused AssetName function

LPGhatguy committed:
Update Changelog

LPGhatguy committed:
Release 0.4.0

LPGhatguy committed:
Replace content-folder target with none.

Closes #16.

LPGhatguy committed:
Upgrade panic handling and default log configuration

LPGhatguy committed:
Migrate from snafu to thiserror

LPGhatguy committed:
Switch to fs-err from custom fs bindings

LPGhatguy committed:
Custom error type for roblox_web_api

LPGhatguy committed:
Gracefully handle sync errors and report them at the end

LPGhatguy committed:
Define conditionals in lua_ast

LPGhatguy committed:
Remove Codegen Configuration (#28)

* Remove codegen configurability

* Change UrlAndSlice codegen to require slice

* Update changelog

LPGhatguy committed:
Implement support for asset-cache-path in tarmac.toml to cache assets

LPGhatguy committed:
Introduce new API for extracting features from DPI-included path

LPGhatguy committed:
Make Expression impl From<&String>

LPGhatguy committed:
Add support for function codegen

LPGhatguy committed:
Clear unknown entries in asset-path-cache before populating it

LPGhatguy committed:
High DPI-aware Tarmac (#31)

* Initial work for high-dpi aware Tarmac

* Fix AST formatting and add helpers

* First pass at high DPI-aware codegen

* Fix failing dpi_scale tests

* Collapse if-let blocks into match

* Introduce concept of paths with DPI information stripped

* Strip file extensions for code generation

LPGhatguy committed:
Add support for uploading group assets (#33)

* Add support for uploading group assets

* Update Changelog

LPGhatguy committed:
Add asset-list-path for generating list of assets

LPGhatguy committed:
Handle HTTP 429 from Roblox sync backend (#34)

LPGhatguy committed:
Update Changelog

LPGhatguy committed:
Flesh out documentation for next release (#35)

* Update documentation to be complete, but terse

* Update README.md

Co-authored-by: Paul Doyle <37384169+ZoteTheMighty@users.noreply.github.com>

Co-authored-by: Paul Doyle <37384169+ZoteTheMighty@users.noreply.github.com>

LPGhatguy committed:
Add Roact example

Roblox does not currently expose resolution scale to user code,
so to emulate its changing, this example has a mock provider
that constantly cycles the scaling factor.

An untested CoreProvider is also included in this example.

Eventually, this code should be turned into a library, but due
to the state of public-facing high DPI aware APIs in the engine,
it's unclear where it would live or how much it would help.

LPGhatguy committed:
Finish unfinished comment

LPGhatguy committed:
Fold includes config into just a list of paths

LPGhatguy committed:
Rename base-path to codegen-base-path on inputs

LPGhatguy committed:
Fix missing update to codegen-base-path

LPGhatguy committed:
Release 0.5.0

LPGhatguy committed:
Change release workflow to be based on tag instead of release creation

LPGhatguy committed:
Automatically retry uploads in tarmac sync if names are moderated

LPGhatguy committed:
Update changelog

LPGhatguy committed:
Add create-cache-map subcommand (#38)

* Add create-path-map subcommand

* Convert to generating cache and map together

* Use root config relative path instead of config relative

* Rename to create-cache-map

* Inline documentation for new command

LPGhatguy committed:
Update Changelog for #38

LPGhatguy committed:
Return error instead of panicking on unknown color spaces

cliffchapmanrbx committed:
Add CLA bot (#40)

LPGhatguy committed:
Migrate from rojo-rbx org to Roblox org

LPGhatguy committed:
Fix remaining references to old Tarmac repo location

LPGhatguy committed:
Implement 'tarmac asset-list' (#41)

* Implement 'tarmac asset-list' to create a list of assets used by the project

* Fix warnings

* Eliminate redundant copy

LPGhatguy committed:
Fix asset-list arg, update docs for it

MisterUncloaked committed:
Release 0.6.0

MisterUncloaked committed:
Merge pull request #42 from MisterUncloaked/0.6.0-release

Release 0.6.0

oltrep committed:
Add retry backend and new command line options

oltrep committed:
fix format

oltrep committed:
remove matches macro

oltrep committed:
remove matches macro in test

oltrep committed:
convert iterator to for loop

oltrep committed:
add log message

oltrep committed:
update readme

ZoteTheMighty committed:
Merge pull request #45 from oltrep/retry-backend

Closes #43

This PR adds a new backend that wraps another one and performs the retry logic when getting rate limited errors.

I've opened this as a draft PR because I have not tested it yet with real image uploads.

oltrep committed:
Release 0.7.0

ZoteTheMighty committed:
Merge pull request #46 from oltrep/release-0.7.0

Release 0.7.0

jeparlefrancais committed:
implement sync to local content folder

jeparlefrancais committed:
remove roblox CLA bot

jeparlefrancais committed:
run cargo fmt

jeparlefrancais committed:
fix clippy warnings

jeparlefrancais committed:
add main branch to ci

jeparlefrancais committed:
Merge pull request #1 from jeparlefrancais/local-sync

implement sync to local content folder

nezuo committed:
Alpha bleed unpacked images (#2)

jeparlefrancais committed:
Release 0.7.1

Vorlias committed:
Fix support for multiple generated sprites in local

If tarmac generates multiple spritesheets, they would conflict with --target local due to having the same sprite names. This changes that.

Vorlias committed:
Include updated ids

Vorlias committed:
Fix?

Vorlias committed:
Fix sync.rs missing alpha_bleed call

Vorlias committed:
Fix roblox auth

Vorlias committed:
Update to rust 1.63.0

Vorlias committed:
Update README.md

Vorlias committed:
Update README.md

Vorlias committed:
Fix SecretString issue

Vorlias committed:
Merge branch 'main' of https://github.com/Vorlias/tarmac

Vorlias committed:
Update README.md

Vorlias committed:
Format

Vorlias committed:
Merge branch 'main' of https://github.com/Vorlias/tarmac

Vorlias committed:
Update Cargo.toml

Vorlias committed:
Fix header woopsie, add CSRF fetch as part of req

Vorlias committed:
Clippy fix

Vorlias committed:
I'm actually being trolled rn

Vorlias committed:
paaaain

Vorlias committed:
Update README.md

Vorlias committed:
update readme

Vorlias committed:
Change to use rojo-rbx repo for installation

Vorlias committed:
Update README.md

Vorlias committed:
Update rbx_cookie to 0.1.4

Vorlias committed:
Merge branch 'main' of https://github.com/rojo-rbx/tarmac

# Conflicts:
#	README.md

kennethloeffler committed:
Add api_key to RobloxApiClient

Use AssetCreator for RobloxSyncBackend.upload_to_group_id + rename

kennethloeffler committed:
Add open cloud implementation

kennethloeffler committed:
Add user ID and group ID options to upload-image command

kennethloeffler committed:
Mention that create-cache-map only works with --auth

kennethloeffler committed:
Pull out API implementations into their own modules + trait

kennethloeffler committed:
Add fallback to legacy when attempting to download with Open Cloud

kennethloeffler committed:
Add upload-to-user-id to readme

kennethloeffler committed:
Add api-key to readme

kennethloeffler committed:
More readme

kennethloeffler committed:
Retrieve Roblox Studio's auth cookie in upload-image command

kennethloeffler committed:
Retrieve Roblox Studio's auth cookie in create_cache_map command

kennethloeffler committed:
Hide api-key's env value in help output

kennethloeffler committed:
Only need one pattern here

kennethloeffler committed:
Reuse GetAsset

kennethloeffler committed:
Create RbxCloud once per session

kennethloeffler committed:
Map from HttpStatusError to ResponseError in From impl

boatbomber committed:
Add config for padding size

boatbomber committed:
Use new padding config

boatbomber committed:
Add to changelog

boatbomber committed:
Add to README

boatbomber committed:
Less clunky wording

boatbomber committed:
Use real PR number now that it's opened

boatbomber committed:
Fix whitespace

Vorlias committed:
Merge pull request #11 from boatbomber/padding-config

Add config for spritesheet padding

Vorlias committed:
Fix #9 - always resync local assets

OutOfBears committed:
Merge remote-tracking branch 'pr/open-cloud' into main

Commit:
start some of the rewrite

Commit:
rip out structopt, use anyhow everywhere for better error handling

RBXTyler commented:
CI / build failed? 👀

Commit:
run cargo fmt everywhere

Commit:
need to add a trait constraint

OutOfBears commented on src/roblox_api/open_cloud.rs:
```diff
@@ -10,83 +13,79 @@ use rbxcloud::rbx::{
-
-    fn upload_image(&mut self, data: &ImageUploadData) -> Result<UploadResponse, RobloxApiError> {
-        self.upload_image_inner(data)
+    // this was a bad idea, sorry
```
Should we remove this?

RBXTyler replied:
We need that method? Because the image will not upload if it is moderated.

OutOfBears replied:
The only reason it ever gets moderated is for name, so we could hash it?

RBXTyler replied:
>:(

RBXTyler replied:
\>:(

OutOfBears commented on src/roblox_api/open_cloud.rs:
```diff
@@ -10,83 +13,79 @@ use rbxcloud::rbx::{
-        LegacyClient::new(self.credentials.clone())?.download_image(id)
+    fn download_image(&self, id: u64) -> Result<Vec<u8>> {
+        todo!();
+        // LegacyClient::new(self.credentials.clone())?.download_image(id)
```
Should we remove this function?

OutOfBears replied:
```suggestion
```

OutOfBears commented on src/sync_backend.rs:
```diff
@@ -32,48 +38,65 @@ pub struct UploadInfo {
+    ApiClient: RobloxApiClient<'a> + Sync + Clone + Send,
+{
+    pub fn new(api_client: ApiClient) -> Self {
+        // Self { api_client: Arc::new(api_client) }
```
\>:(

OutOfBears replied:
```suggestion
```

OutOfBears commented on src/sync_backend.rs:
```diff
@@ -32,48 +38,65 @@ pub struct UploadInfo {
-        }
+        todo!()
+
+        // let result = self
```
\>:(

OutOfBears commented on src/sync_backend.rs:
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
-        Ok(UploadResponse {
-            id: AssetId::Id(id),
-        })
+        // Ok(UploadResponse {
```
\>:(

OutOfBears commented on src/sync_backend.rs:
```diff
@@ -234,116 +264,116 @@ pub enum Error {
-        }
-    }
-}
+// #[cfg(test)]
```
\>:(

OutOfBears commented on src/main.rs:
```diff
@@ -14,26 +14,31 @@ mod sync_backend;
-        Subcommand::Sync(sync_options) => commands::sync(options.global, sync_options)?,
-        Subcommand::CreateCacheMap(sub_options) => {
+        Command::Sync(sync_options) => {
+            // commands::sync(options.global, sync_options)?,
```
```suggestion
```

OutOfBears commented on src/roblox_api/mod.rs:
```diff
@@ -33,19 +34,19 @@ pub struct RobloxCredentials {
+pub trait RobloxApiClient<'a> {
+    fn new(credentials: RobloxCredentials) -> Result<Self>
+    where Self: Sized;
+    // this was a bad idea, sorry
+    // async fn upload_image_with_moderation_retry(
+    //     &self,
+    //     data: ImageUploadData<'a>,
+    // ) -> Result<UploadResponse>;
```
```suggestion
```

OutOfBears commented on src/roblox_api/mod.rs:
```diff
@@ -129,7 +130,8 @@ Tarmac will attempt to upload to the user currently logged into Roblox Studio, o
 
-            Ok(Box::new(LegacyClient::new(credentials)?))
+            todo!();
+            // Ok(Box::new(LegacyClient::new(credentials)?))
```
```suggestion
```

OutOfBears commented on src/roblox_api/open_cloud.rs:
```diff
@@ -10,83 +13,79 @@ use rbxcloud::rbx::{
-
-    fn upload_image(&mut self, data: &ImageUploadData) -> Result<UploadResponse, RobloxApiError> {
-        self.upload_image_inner(data)
+    // this was a bad idea, sorry
+    // async fn upload_image_with_moderation_retry(
+    //     &self,
+    //     data: ImageUploadData<'a>,
+    // ) -> Result<UploadResponse> {
+    //     match self.upload_image(data.clone()).await {
+    //         Err(RobloxApiError::ResponseError { status, body })
+    //             if status == 400 && body.contains("moderated") =>
+    //         {
+    //             log::warn!(
+    //                 "Image name '{}' was moderated, retrying with different name...",
+    //                 data.name
+    //             );
+    //             self.upload_image(ImageUploadData {
+    //                 name: "image".to_string(),
+    //                 ..data.to_owned()
+    //             })
+    //             .await
+    //         }
+
+    //         result => result,
+    //     }
+    // }
```
```suggestion
```

OutOfBears commented on src/sync_backend.rs:
```diff
@@ -32,48 +38,65 @@ pub struct UploadInfo {
-        }
+        todo!()
+
+        // let result = self
+        //     .api_client
+        //     .upload_image_with_moderation_retry(&ImageUploadData {
+        //         image_data: Cow::Owned(data.contents),
+        //         name: &data.name,
+        //         description: "Uploaded by Tarmac.",
+        //     }).await;
+
+        // match result {
+        //     Ok(response) => {
+        //         log::info!(
+        //             "Uploaded {} to ID {}",
+        //             &data.name,
+        //             response.backing_asset_id
+        //         );
+
+        //         Ok(UploadResponse {
+        //             id: AssetId::Id(response.backing_asset_id),
+        //         })
+        //     }
+
+        //     Err(RobloxApiError::ResponseError {
+        //         status: StatusCode::TOO_MANY_REQUESTS,
+        //         ..
+        //     }) => Err(Error::RateLimited),
+
+        //     Err(err) => Err(err.into()),
+        // }
```
```suggestion
```

OutOfBears commented on src/sync_backend.rs:
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
-        log::info!("Copying {} to local folder", &data.name);
+    async fn upload(&self, data: UploadInfo) -> Result<UploadResponse, Error> {
+        todo!();
+        // log::info!("Copying {} to local folder", &data.name);
```
```suggestion
```

OutOfBears commented on src/sync_backend.rs:
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
 
-        self.last_id += 1;
-        let id = self.last_id;
+        // self.last_id += 1;
+        // let id = self.last_id;
```
```suggestion

```

OutOfBears commented on src/sync_backend.rs:
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
 
-        let path = Path::new(".tarmac-debug");
-        fs::create_dir_all(path)?;
+        // let path = Path::new(".tarmac-debug");
+        // fs::create_dir_all(path)?;
```
```suggestion

```

OutOfBears commented on src/sync_backend.rs:
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
 
-        let file_path = path.join(id.to_string());
-        fs::write(&file_path, &data.contents)?;
+        // let file_path = path.join(id.to_string());
+        // fs::write(&file_path, &data.contents)?;
```
```suggestion

```

OutOfBears commented on src/sync_backend.rs:
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
-        Ok(UploadResponse {
-            id: AssetId::Id(id),
-        })
+        // Ok(UploadResponse {
+        //     id: AssetId::Id(id),
+        // })
```
```suggestion

```

OutOfBears commented on src/sync_backend.rs:
```diff
@@ -234,116 +264,116 @@ pub enum Error {
+//         #[test]
+//         fn upload_returns_first_success_result() {
+//             let mut counter = 0;
+//             let success = UploadResponse {
+//                 id: AssetId::Id(10),
+//             };
+//             let inner = CountUploads::new(&mut counter).with_results(vec![
+//                 Err(Error::RateLimited),
+//                 Err(Error::RateLimited),
+//                 Ok(success.clone()),
+//             ]);
+//             let mut backend = RetryBackend::new(inner, 5, retry_duration());
+
+//             let upload_result = backend.upload(any_upload_info()).unwrap();
+
+//             assert_eq!(counter, 3);
+//             assert_eq!(upload_result, success);
+//         }
+
+//         #[test]
+//         fn upload_returns_rate_limited_when_retries_exhausted() {
+//             let mut counter = 0;
+//             let inner = CountUploads::new(&mut counter).with_results(vec![
+//                 Err(Error::RateLimited),
+//                 Err(Error::RateLimited),
+//                 Err(Error::RateLimited),
+//                 Err(Error::RateLimited),
+//             ]);
+//             let mut backend = RetryBackend::new(inner, 2, retry_duration());
+
+//             let upload_result = backend.upload(any_upload_info()).unwrap_err();
+
+//             assert_eq!(counter, 3);
+//             assert!(match upload_result {
+//                 Error::RateLimited => true,
+//                 _ => false,
+//             });
+//         }
+//     }
+// }
```
```suggestion

```

OutOfBears commented on src/sync_backend.rs:
```diff
@@ -143,28 +168,32 @@ impl DebugSyncBackend {
 /// Performs the retry logic for rate limitation errors. The struct wraps a SyncBackend so that
 /// when a RateLimited error occurs, the thread sleeps for a moment and then tries to reupload the
 /// data.
+///
```
```suggestion
```

RBXTyler commented on src/roblox_api/open_cloud.rs:
```diff
@@ -10,83 +13,79 @@ use rbxcloud::rbx::{
-    fn download_image(&mut self, id: u64) -> Result<Vec<u8>, RobloxApiError> {
-        LegacyClient::new(self.credentials.clone())?.download_image(id)
+    fn download_image(&self, id: u64) -> Result<Vec<u8>> {
+        todo!();
```
????????????

Commit:
Refactor code and add async support

***Add resize option to image upload***

***Update error handling in create_cache_map***

***Fix missing imports and unused code***

Commit:
Slightly refactor image upload function and update asset ID logging

Commit:
more refactoring

Commit:
move command options into their own modules, rename GlobalOptions -> Global

Commit:
remove debug print, use logger for info print

Commit:
run cargo fmt on everything

Commit:
re-enable downloading images from open cloud client
